File: app/src/main/java/io/github/shadow578/music_dl/backup/BackupGSONAdapters.java
package io.github.shadow578.music_dl.backup;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

import io.github.shadow578.music_dl.KtPorted;

/**
 * gson adapters for backup data
 */
@KtPorted
public class BackupGSONAdapters {

    @KtPorted
    public static class LocalDateTimeAdapter extends TypeAdapter<LocalDateTime> {

        private static final DateTimeFormatter FORMAT = DateTimeFormatter.ISO_DATE_TIME;

        @Override
        public void write(JsonWriter out, LocalDateTime value) throws IOException {
            if (value == null) {
                out.nullValue();
            } else {
                out.value(FORMAT.format(value));
            }
        }

        @Override
        public LocalDateTime read(JsonReader in) throws IOException {
            if (in.peek().equals(JsonToken.NULL)) {
                in.nextNull();
                return null;
            } else {
                return LocalDateTime.parse(in.nextString(), FORMAT);
            }
        }
    }

    @KtPorted
    public static class LocalDateAdapter extends TypeAdapter<LocalDate> {

        private static final DateTimeFormatter FORMAT = DateTimeFormatter.ISO_DATE;

        @Override
        public void write(JsonWriter out, LocalDate value) throws IOException {
            if (value == null) {
                out.nullValue();
            } else {
                out.value(FORMAT.format(value));
            }
        }

        @Override
        public LocalDate read(JsonReader in) throws IOException {
            if (in.peek().equals(JsonToken.NULL)) {
                in.nextNull();
                return null;
            } else {
                return LocalDate.parse(in.nextString(), FORMAT);
            }
        }
    }
}

File: app/src/main/kotlin/io/github/shadow578/yodel/backup/BackupGSONAdapters.kt
package io.github.shadow578.yodel.backup

import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonToken
import com.google.gson.stream.JsonWriter
import java.io.IOException
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * gson adapter for [LocalDateTime]
 */
class LocalDateTimeAdapter : TypeAdapter<LocalDateTime?>() {
    @Throws(IOException::class)
    override fun write(writer: JsonWriter, value: LocalDateTime?) {
        if (value == null) {
            writer.nullValue()
        } else {
            writer.value(FORMAT.format(value))
        }
    }

    @Throws(IOException::class)
    override fun read(reader: JsonReader): LocalDateTime? {
        return if (reader.peek() == JsonToken.NULL) {
            reader.nextNull()
            null
        } else {
            LocalDateTime.parse(reader.nextString(), FORMAT)
        }
    }

    companion object {
        private val FORMAT = DateTimeFormatter.ISO_DATE_TIME
    }
}

/**
 * gson adapter for [LocalDate]
 */
class LocalDateAdapter : TypeAdapter<LocalDate?>() {
    @Throws(IOException::class)
    override fun write(writer: JsonWriter, value: LocalDate?) {
        if (value == null) {
            writer.nullValue()
        } else {
            writer.value(FORMAT.format(value))
        }
    }

    @Throws(IOException::class)
    override fun read(reader: JsonReader): LocalDate? {
        return if (reader.peek() == JsonToken.NULL) {
            reader.nextNull()
            null
        } else {
            LocalDate.parse(reader.nextString(), FORMAT)
        }
    }

    companion object {
        private val FORMAT = DateTimeFormatter.ISO_DATE
    }
}